mlflow/go/pkg/tracking/store/sql/runs_test.go

```diff
@@ -0,0 +1,239 @@
+	{
+		name:  "simple metric query",
+		query: "metrics.accuracy > 0.72",
+		expectedSQL: `
+SELECT "run_uuid" FROM "runs"
+JOIN (SELECT "run_uuid","value" FROM "latest_metrics" WHERE key = $1 AND value > $2)
+AS filter_0
+ON runs.run_uuid = filter_0.run_uuid`,
+		expectedVars: []any{"accuracy", 0.72},
+	},
+	{
+		name:  "simple metric and param query",
+		query: "metrics.accuracy > 0.72 AND params.batch_size = '2'",
+		expectedSQL: `
+SELECT "run_uuid" FROM "runs"
+JOIN (SELECT "run_uuid","value" FROM "latest_metrics" WHERE key = $1 AND value > $2)
+AS filter_0 ON runs.run_uuid = filter_0.run_uuid
+JOIN (SELECT "run_uuid","value" FROM "params" WHERE key = $3 AND value = $4)
+AS filter_1 ON runs.run_uuid = filter_1.run_uuid
+`,
+		expectedVars: []any{"accuracy", 0.72, "batch_size", "2"},
+	},
+	{
+		name:  "tag query",
+		query: "tags.environment = 'notebook' AND tags.task ILIKE 'classif%'",
+		expectedSQL: `
+SELECT "run_uuid" FROM "runs"
+JOIN (SELECT "run_uuid","value" FROM "tags" WHERE key = $1 AND value = $2)
+AS filter_0 ON runs.run_uuid = filter_0.run_uuid
+JOIN (SELECT "run_uuid","value" FROM "tags" WHERE key = $3 AND value ILIKE $4)
+AS filter_1 ON runs.run_uuid = filter_1.run_uuid`,
+		expectedVars: []any{"environment", "notebook", "task", "classif%"},
+	},
+	{
+		name:  "datasests IN query",
+		query: "datasets.digest IN ('s8ds293b', 'jks834s2')",
+		expectedSQL: `
+SELECT "run_uuid" FROM "runs"
+JOIN (SELECT "experiment_id","digest" FROM "datasets" WHERE digest IN ($1,$2))
+AS filter_0 ON runs.experiment_id = filter_0.experiment_id
+`,
+		expectedVars: []any{"s8ds293b", "jks834s2"},
+	},
+	{
+		name:  "attributes query",
+		query: "attributes.run_id = 'a1b2c3d4'",
+		expectedSQL: `
+SELECT "run_uuid" FROM "runs"
+WHERE runs.run_uuid = $1
+`,
+		expectedVars: []any{"a1b2c3d4"},
+	},
+	{
+		name:  "run_name query",
+		query: "attributes.run_name = 'my-run'",
+		expectedSQL: `
+SELECT "run_uuid" FROM "runs"
+JOIN (SELECT "run_uuid","value" FROM "tags" WHERE key = $1 AND value = $2)
+AS filter_0 ON runs.run_uuid = filter_0.run_uuid`,
+		expectedVars: []any{"mlflow.runName", "my-run"},
+	},
+	{
+		name:  "datasets.context query",
+		query: "datasets.context = 'train'",
+		expectedSQL: `
+SELECT "run_uuid" FROM "runs"
+JOIN (
+	SELECT inputs.destination_id AS run_uuid
+	FROM "inputs"
+	JOIN input_tags
+	ON inputs.input_uuid = input_tags.input_uuid
+	AND input_tags.name = 'mlflow.data.context'
+	AND input_tags.value = $1
+	WHERE inputs.destination_type = 'RUN'
+) AS filter_0 ON runs.run_uuid = filter_0.run_uuid`,
+		expectedVars: []any{"train"},
+	},
+}
+
+func assertTestData(t *testing.T, database *gorm.DB, testData testData) {
+	t.Helper()
+
+	transaction := database.Model(&models.Run{})
+
+	contractErr := sql.ApplyFilter(database, transaction, testData.query)
+	if contractErr != nil {
+		t.Fatal("contractErr: ", contractErr)
+	}
+
+	actualSQL := transaction.Select("ID").Find(&models.Run{}).Statement.SQL.String()
```

Comment:
it's a bit weird that you finalise the transaction _and_ assign in one go here, and then use `transaction.Statement.Vars` further
I'd be more inclined to just finalise the transaction, check whether any errors were encountered, and then use `transaction.Statement.SQL.String()` directly in the assertion